feat: a weaker definition of compactly generated spaces

scholzhannah opened the pull request:
This PR defines compactly generated spaces as a topological space where a set is open iff its intersection with every compact set is open in that compact set. There is a [definition of compactly generated spaces already in mathlib](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Compactness/CompactlyGeneratedSpace.html#CompactlyGeneratedSpace) but that definition is stronger than the one presented here. They do however agree on Hausdorff spaces. See [this wikipedia page](https://en.wikipedia.org/wiki/Compactly_generated_space) for an explanation for the three different definitions of compactly generated spaces used in literature. 

Co-authored-by: Floris van Doorn <fpvdoorn@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

- [ ] depends on: #21134

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

scholzhannah commented:
I am aware that the naming is not ideal since in the literature "k-space" and "compactly generated space" are used as synonyms. Better ideas are very much welcome. I personally like the name "k-space" because then you can use "k-ification" for the operation of turning an arbitrary topological space into a k-space. I don't think there is a good name for this operation when using "compactly generated space" instead of "k-space".

jzxia commented:
Hello @scholzhannah, thank you for your work!  Would you mind if I open another PR that depends on this one to prove the exponential law? It states that if $X$, $Y$, and $Z$ are compactly generated Hausdorff spaces, then $kC(Y \times X, Z)$ is homeomorphic to $kC(X, kC(Y, Z))$, where $k$ is your k-ification and $C(X, Y)$ denotes the space of continuous maps from $X$ to $Y$, with the compact-open topology. The exponential law is important for the homotopy theory of CW-complexes. In particular, for the Whitehead theorem, I need this law to prove that every relative CW-complex has the homotopy extension property.

scholzhannah commented:
> Would you mind if I open another PR that depends on this one to prove the exponential law?

Sure, of course I don't mind!

jzxia commented:
> open another PR that depends on this one to prove the exponential law

I ended up only proving the exponential law $\text{Hom}(Y \times X, Z) \equiv \text{Hom}(Y, C(X,Z))$ for locally compact $X$ which seems to be adequate for the Whitehead theorem. I might come back when I have the time to refactor and generalize my code, but if you (or anyone else) want to prove the general exponential law, please feel free to do so.

urkud commented:
@StevenClontz WDYT about naming?
@scholzhannah Note that `RestrictGenTopology` will very soon be renamed to `IsCoherentWith`, see #18397

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+
+/-- If every set `A` is open if for every compact `K` the intersection `K ∩ A` is open in `K`,
+then the space is a k-space. -/
+lemma kspace_of_isOpen {X : Type*} [TopologicalSpace X]
```
Why not `of_isOpen`? The full name will be `KSpace.of_isOpen`, which IMHO makes sense. Also, please reuse `X` from `variable`s.

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+  restrictGenTopology := RestrictGenTopology.of_isClosed h
+
+/-- Every weakly locally compact space is a k-space.-/
+instance kspace_of_WeaklyLocallyCompactSpace {X : Type*} [TopologicalSpace X]
```
```suggestion
instance of_weaklyLocallyCompactSpace
```

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+  restrictGenTopology := RestrictGenTopology.isCompact_of_weaklyLocallyCompact
+
+/-- Every sequential space is a k-space.-/
+instance kspace_of_SequentialSpace {X : Type*} [TopologicalSpace X]
```
```suggestion
instance of_sequentialSpace
```

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+
+/-- In a k-space `X`, a set `s` is open when `f ⁻¹' s` is open for every continuous map from a
+compact space. -/
+lemma preimage_isOpen.{u} {X : Type u} [TopologicalSpace X] [KSpace X] (s : Set X)
```
You can use
```
universe u

variable {X : Type u}
```
above, then you can reuse `X` here.

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+
+/--For a topological space `X` the k-ification is defined as:
+`A` is open iff for all compact sets `B`, the intersection `A ∩ B` is open in `B`.-/
+instance instkification {X : Type*} [TopologicalSpace X] : TopologicalSpace (kification X) where
```
```suggestion
instance kification.instTopologicalSpace : TopologicalSpace (kification X) where
```

urkud commented on Mathlib/Topology/Compactness/KSpace.lean:
```diff
@@ -0,0 +1,315 @@
+/-- A set `A` is the k-ification is open iff for all compact sets `K`,
+the intersection `K ∩ A` is open in `K`.-/
+lemma kification.isOpen_iff {X : Type*} [TopologicalSpace X] {A : Set (kification X)} :
+    IsOpen (X := kification X) A ↔
```
Please don't abuse defeq `kification X = X`. E.g., `simp` in Lean 4 is very unhappy about this type of abuse. Instead, please define
```
def toKification : X ≃ kification X := Equiv.refl _
```
then use images or preimages under `toKification` and `toKification.symm` to move sets between the spaces. This also applied to the `TopologicalSpace` instance.

urkud commented:
I left style comments on about half of the file. Please take them into accounts, then remove the `awaiting-author` label.

StevenClontz commented:
Yeah this ambiguity (without the [weak] Hausdorff assumption) is why there are three distinct variations for "k-space" in π-Base...

https://topology.pi-base.org/properties?filter=K-space

One might say a k1-space (which is what I understand this PR proposes) is "coherent with its compact subspaces".

I'd need to spend more time with this to suggest something clever. But it would be unfortunate to have "compactly generated" and "k-space" represent inequivalent concepts in mathlib..

StevenClontz commented:
One possible naming convention, should we want to lean in on using the term "coherence" consistently when discussing spaces generated by subspaces.

- `CompactlyCoherentSpace X`: this PR, implementing https://topology.pi-base.org/properties/P000140
- `CompactlyGeneratedSpace X`: current property in mathlib implementing https://topology.pi-base.org/properties/P000141 (I don't love this as such spaces are really generated by *continuous maps* from *Hausdorff* compacta - I'd rather this be `KSpace X` actually, but it's certainly out of scope for this PR to make such a change, so sticking with status quo makes sense to me.)
- `T2CompactlyCoherentSpace X`: possible future contribution implementing https://topology.pi-base.org/properties/P000142

Or I could see renaming all as `K1Space X`, `K2Space X`, `K3Space X` perhaps. But as I said, it will be surprising to have inequivalent properties in mathlib named `KSpace X` and `CompactlyGeneratedSpace X`.

urkud commented:
@scholzhannah WDYT about the naming proposed by @StevenClontz ?

scholzhannah commented:
Hm, I am not sure actually. I think for myself I would like `K1Space`, `K2Space` and `K3Space` because their similar names make it clear how closely the concepts are related. Additionally,  I could then call what is now the `kification` the `k1ification` which would still make sense. But for mathlib I think that soultion obscures the concepts behind non-standard naming, so I am not sure that this is the way to go. `CompactlyCoherentSpace`, `CompactlyGeneratedSpace` and `T2CompactlyCoherentSpace` feel more mathlib style to me anyways.

StevenClontz commented:
Yeah, the $k_n$ scheme was something we more or less made up for pi-Base. It *might* end up in the literature if a preprint or two of mine are eventually accepted, but it's certainly not standard. I think Coherent/Generated is the more semantic and mathlib-y way to go as well.

scholzhannah commented:
Okay, then I will get on renaming `KSpace` to `CompactlyCoherentSpace` soon.

ADedecker commented:
I had noticed this a few weeks ago but didn't have time to check it out. I'm also interested in this definition for application to proper actions, where the current "Compactly Generated" in Mathlib forces slightly stronger assumption (IIRC). I was a bit sad about "losing" the naming race to category theorists, but I love `CompactlyCoherentSpace`, so thank you for adding it and thanks to @StevenClontz for figuring out the name!

scholzhannah commented:
I have renamed `KSpace` to `CompactlyCoherentSpace`. Additionally, I renamed `kification` to `CompactCoherentification` and the map `toKification` to `toCompactlyCoherent` for now. I am not sure if those are good names. Especially because I am not aware of the naming conventions for type synonyms. I have chosen `CompactCoherentification` because from what I understand the "coherentification" is a process that can be done with any family of subsets of a topological space. (So in my mind the word "coherentification" is somewhat of a necessary evil.) And this is just the compact version of that.

ADedecker commented:
Could you update the PR description to the new name ?

ADedecker commented on Mathlib/Topology/Compactness/CompactlyCoherentSpace.lean:
```diff
@@ -0,0 +1,336 @@
+    CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.of_isClosed h
+
+/-- Every weakly locally compact space is a compactly coherent space. -/
+instance of_weaklyLocallyCompactSpace [WeaklyLocallyCompactSpace X] : CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.isCompact_of_weaklyLocallyCompact
+
+/-- Every sequential space is a compactly coherent space. -/
+instance of_sequentialSpace [SequentialSpace X] : CompactlyCoherentSpace X where
+  isCoherentWith := IsCoherentWith.isCompact_of_seq
```
Could you instead deprecate the existing results? Or, if that is too much work, at least add a TODO somewhere saying that every lemma using `IsCoherentWith` with compact subspaces should b updated to use `CompactlyCoherentSpace`?

ADedecker commented on Mathlib/Topology/Compactness/CompactlyCoherentSpace.lean:
```diff
@@ -0,0 +1,336 @@
+
+/-- A set `A` in a compactly coherent space is open iff for every compact set `K`, the intersection
+`K ∩ A` is open in `K`. -/
+lemma isOpen_iff [CompactlyCoherentSpace X] (A : Set X) :
```
I think `A` should be made implicit here, for consistency with the rest of the library.

ADedecker commented on Mathlib/Topology/Compactness/CompactlyCoherentSpace.lean:
```diff
@@ -0,0 +1,336 @@
+
+/-- In a compactly coherent space `X`, a set `s` is open when `f ⁻¹' s` is open for every continuous
+map from a compact space. -/
+lemma preimage_isOpen [CompactlyCoherentSpace X] (s : Set X)
+    (hs : (∀ (K : Type u) [TopologicalSpace K],
+      [CompactSpace K] → ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s))) :
+    IsOpen s := by
+  rw [isOpen_iff]
+  intro K hK
+  let _ : CompactSpace (Elem K) := isCompact_iff_compactSpace.mp hK
+  exact hs (Elem K) Subtype.val continuous_subtype_val
```
I would upgrade this to an iff:
```suggestion
lemma isOpen_iff_forall_compactSpace [CompactlyCoherentSpace X] (s : Set X) : 
    IsOpen s ↔ 
      ∀ (K : Type u) [TopologicalSpace K] [CompactSpace K], 
      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s) := by
  refine ⟨fun hs _ _ _ _ hf ↦ hs.preimage hf, fun hs ↦ isOpen_iff _ |>.mpr ?_⟩
  intro K hK
  let _ : CompactSpace K := isCompact_iff_compactSpace.mp hK
  exact hs K Subtype.val continuous_subtype_val
```
(We don't usually spell out `Elem`, instead we let the coercion mechanism do its magic)

ADedecker commented on Mathlib/Topology/Compactness/CompactlyCoherentSpace.lean:
```diff
@@ -0,0 +1,336 @@
+
+/-- A topological space `X` is compactly coherent if a set `s` is open when `f ⁻¹' s?` is open for
+every continuous map `f : K → X`, where `K` is compact. -/
+lemma of_preimage_isOpen (h : ∀ (s : Set X), (∀ (K : Type u) [TopologicalSpace K],
+      [CompactSpace K] → ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) → IsOpen s) :
+    CompactlyCoherentSpace X where
+  isCoherentWith := {
+    isOpen_of_forall_induced := by
+      intro s hs
+      apply h s
+      intro K _ _ f hf
+      change IsOpen ((fun x ↦ ⟨f x, mem_image_of_mem f trivial⟩) ⁻¹' ((f '' univ) ↓∩ s))
+      apply (hf.subtype_mk fun x ↦ mem_image_of_mem f trivial).isOpen_preimage
+      exact hs (f '' univ) (CompactSpace.isCompact_univ.image hf)}
```
```suggestion
lemma of_isOpen_forall_compactSpace (h : ∀ (s : Set X), (∀ (K : Type u) [TopologicalSpace K],
      [CompactSpace K] → ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) → IsOpen s) :
    CompactlyCoherentSpace X := by
  refine of_isOpen fun A hA ↦ h A fun K _ _ f hf ↦ ?_
  specialize hA (range f) (isCompact_range hf)
  have := hA.preimage (hf.codRestrict mem_range_self)
  rwa [← preimage_comp] at this
```
(Apart from the golf, I'm also suggesting another name, feel free to disagree)

ADedecker reviewed (commented):
(More comments incoming)

scholzhannah commented:
> With the comments above, I believe the first part of the file is ready. I played a bit with the k-ification section to simplify it, and I ended up pushing to branch "AD_kification".

Great, thanks a lot! I'll make a dependent PR for the second part of the file.
